### ISSUE TITLE
Renaming MagicDNS URLs to. machine name only.

### ISSUE THREAD

**Issue by darshinimashar** [tailscale/tailscale#1543]
Also, for DNS, the Customer would really like to make DNS aliases such as **example.ourdomain.beta.tailscale.net** always point to a machine name **staging-1** or something like that. 

This way, they can easily access the machines.

**Comment by guidoiaquinti** [comment]
I would love to see as well support for an internal `.tailscale` domain name (example: `nas.tailscale`, `pi.tailscale`) and maybe even subzones (example: `nas.home.tailscale` VS `nas.office.tailscale`) without the need of a dedicated/external DNS server.

**Comment by MingSpace** [comment]
This feature will easily customize the DNS of the home server , just use tailscale 😀

> I would love to see as well support for an internal `.tailscale` domain name (example: `nas.tailscale`, `pi.tailscale`) and maybe even subzones (example: `nas.home.tailscale` VS `nas.office.tailscale`) without the need of a dedicated/external DNS server.

**Comment by bradfitz** [comment]
0debb99f0852e8fa53ab6c725966f8dd99b3fbc9 added client support for a new ExtraRecords field. It's not yet supported server-side.

**Comment by awoimbee** [comment]
Sorry to be this person, but: Any news on this @bradfitz ? It's awkward that it has been partially implemented for 7 months.
This would simplify the process described here https://github.com/tailscale/tailscale/issues/1748#issuecomment-824121315 quite a bit by removing the need to host a DNS server.

**Comment by bradfitz** [comment]
> It's awkward that it has been partially implemented for 7 months.

Well, it was "partially implemented" because we needed the part we implemented (and now use) for the LetsEncrypt support.

Exposing more of it for this has a ton of product & security considerations and interacts often surprisingly with various DNS stacks on different platforms, which will increase our support burden if we ship something that seems to work sometimes but not on all platforms in all configs.

Example of problems:

* if you add a `foo.net` record, we need to make sure your OS queries the 100.100.100.100 resolver for that record. But some OSes only let us route `*.foo.net` to us, which means what do we do with `bar.foo.net`? We'd have to forward that on upstream, but some OSes don't make it easy or possible for us to figure out what the system's underlying/native DNS servers are. So then we'd require you specify an upstream DNS server in your Tailnet config, which we're trying to move away from so we can enable MagicDNS on by default.
* if you add a `foo.tailnet-name.ts.net` record, what if `foo` already exists as a machine name on the tailnet? what if `foo` comes into existence later?  does that reserve `foo` for you? do we permit ACME challenges for `foo`? what if its IP is changed later? which node can then do the ACME challenge?
* if you add a record pointing at the IP of a node and share that node, does the extra record get shared to other tailnets? (not for now, but that's a possible use case people might want later)

We're leaning towards shipping this soon but with a bunch of limitations to start like:
* no *.ts.net
* no *.tailscale.net
* must have DNS servers set with force override (unfortunate for now, but makes things consistent)

Then we probably won't box ourselves into a corner and have to remove functionality later and can add more of this when other parts are ready.

The real problem here is that this feature means a number of things to different people and we have to make sure we don't try to do all of it at once, poorly.

**Comment by Tener** [comment]
@bradfitz any news about the release date for the limited version of this feature? Having Tailscale-integrated DNS is a great feature, but a little bit of flexibility would be very welcome.

**Comment by DentonGentry** [comment]
- I promise we will update this bug when something is close to being ready
- polling for updates is not useful

**Comment by Tener** [comment]
Thanks, I didn't mean to push, merely hoping to learn of your plans with this feature.

**Comment by mpl** [comment]
@DentonGentry Hello. I understand the meaning of the P2 and T0 labels, but I'm not sure I get the L1 one. Does it describe the likelihood of the issue to get fixed? oh, or does it give an indication of the number of other issues similar to this one?

And I have a question about the feature itself that would address this issue:
Do you envision that there will be a public API for adding DNS entries (which means it could be done with the CLI maybe), or will the feature be limited to the admin web interface?

**Comment by DentonGentry** [comment]
It is an estimate of the number of sites which would use this feature request. Given the number of 👍 since the label was applied, I changed the assessment.

I expect an initial way to make the functionality available will be a new field in the policy JSON in https://login.tailscale.com/admin/acls. There is an API for that: https://github.com/tailscale/tailscale/blob/main/api.md#tailnet-acl-post

**Comment by hybras** [comment]
I'm currently have jobs on my machine's that get my `tailscale ip` and set dns records with my dynamic dns provider. I then have CNAMEs to these records. I get https certificates using let's encrypt's DNS challenge because these services are only available from my network / tailscale. This works beautifully though it was a pain to figure out. 

If tailscale's dns supported arbitrary records, I could cut out the ddns middle man since tailscale already has records pointing to my nodes.

**Comment by schmiman** [comment]
Following, this would really help our tailscale adoption.

**Comment by danjl1100** [comment]
Just thinking out loud for the "means many things to different people" aspect notes above, a possible minimal first step is to only allow subdomains for an existing host on the tailnet.

E.g.
For a host assigned: `foo.tailnet-name.ts.net`

You could configure specific subdomains underneath that host:
```text
sub1.foo.tailnet-name.ts.net
sub2.foo.tailnet-name.ts.net
```

Unclear how big of a leap `*.foo.tailnet-name.ts.net` would be (technically/security-wise) but just a distinct list of subdomains per machine would be a good starting point.
To me, having higher level subdomains `sub-higher.tailnet-name.ts.net` doesn't need to be in the initial offering.

**Comment by FiloSottile** [comment]
As an interesting prior art to look at, there's dnscrypt_proxy's cloaking feature, which is what I am using at the moment and that I'd hope to replace with this. https://github.com/DNSCrypt/dnscrypt-proxy/wiki/Cloaking

**Comment by kai-tub** [comment]
Sorry for piling on, but I tried to get my local server up and running with DNSCrypt + tailscale and failed miserably.
I _felt_ the blog post from tailscale [The Sisyphean Task Of DNS Client Config on Linux](https://tailscale.com/blog/sisyphean-dns-client-linux/) and fought so loooong with systemd-resolved/Networkmanager/DNScrypt etc. that I gave up and looked for an alternative.
One that is [offically integrated](https://tailscale.com/kb/1218/nextdns/) with tailscale is [NextDNS](https://nextdns.io/), which also supports this feature. There it is called "Rewrites" under the "Settings" menu.
That takes about 30min to set up and you are potentially saving yourself _many_ frustrating evenings. :sweat_smile: 

The main drawback of this approach is: You cannot get proper HTTPS support for the tailscale subdomains, as MagicDNS will kill off trying to access any subdomains...
I would like to have at least the option to query the next DNS server in line (NextDNS) that would point right back to the original. But I am very new to DNS and automatic certificate generation, so maybe there is a good reason not to do it.

Hope it helps someone :)

**Comment by artis3n** [comment]
Hope it is not obnoxious to specify a use case for this I don't really see yet described in the comments - I'd love to register subdomains of a host in my tailnet as @danjl1100 describes above to basically solve https://caddy.community/t/the-subfolder-problem-or-why-cant-i-reverse-proxy-my-app-into-a-subfolder/8575 when following the pattern from https://tailscale.com/kb/1166/vscode-ipad/ for other purposes (in my case, making HashiCorp Vault run inside a tailnet with more than one thing running on that host).

**Comment by casidiablo** [comment]
I also tried using NextDNS integration to solve the subdomain issue, and failed miserably.

In the nextdns context: why can't we use that as the resolver for everything and then fall back to use 100.100.100.100 for tailnet domains?

**Comment by trollybusman** [comment]
Here is how I solved a similar issue using a service kind of like NextDNS, it required additional software but it runs seamlessly in the background. 

1. Get Control D's proxy daemon: https://github.com/Control-D-Inc/ctrld
2. Use a config like this:

```
[upstream.0]
    bootstrap_ip = "76.76.2.11"
    endpoint = "https://dns.controld.com/REDACTED"
    name = "Control D - Main org"
    timeout = 2000
    type = "doh"

[upstream.1]
    endpoint = "100.100.100.100"
    name = "Tailscale"
    timeout = 2000
    type = "legacy"

[network.0]
    cidrs = ["0.0.0.0/0"]
    name = "Network 0"
	
[listener.0]
    ip = "127.0.0.1"
    port = 53
    restricted = false

	[listener.0.policy]
	name = "Internal domains"
	rules = [
		{"*.int.company.com" = ["upstream.1"]},
		{"*.ts.net" = ["upstream.1"]},
	]
	```

3. Start the above in service mode: `ctrld start --config /path/to/config`

All internal domains get routed to Tailscale DNS, while everything else uses my Control D resolver. If issue #7946 gets done, can probably skip using the software.

**Comment by nhumrich** [comment]
Just adding my two cents in case it helps. I would like this feature. My use case is I want multiple dns records to point to a single machine, basically, aliases for machines. The reason is, I have a single machine that is hosting multiple things. I can use a reverse proxy such as caddy to proxy to different apps based on domain. Without this feature, I have to use public cnames, or host my own internal dns server.

**Comment by pdone** [comment]
I would like this feature.

**Comment by ndlarsen** [comment]
> Just adding my two cents in case it helps. I would like this feature. My use case is I want multiple dns records to point to a single machine, basically, aliases for machines. The reason is, I have a single machine that is hosting multiple things. I can use a reverse proxy such as caddy to proxy to different apps based on domain. Without this feature, I have to use public cnames, or host my own internal dns server.

My usecase is similar. Running an nginx reverse proxy in front of multiple containerized applications including vaultwarden. I would like to be able to reach the vaultwarden instance via the reverse proxy on the same domain regardless of being connected to the tailscale based network or the actual network containing the proxy. So, yeah, +1 for this feature.

**Comment by pw-64** [comment]
My usecase:
Traefik container serves the other containers in the docker-compose through `containerName.myserver.address`. So this makes it really easy so I don't have to remember ports. I can just go to `dashy.myserver.localhost` and it works great (localhost while testing, going to have a pihole running soon so I can use a dns thing because it is not possible to do `http://containerName.myserver.IP_address/`). However, since I can only access `http://tailscale_device_name/` from another device over tailscale, I cannot go to the subdomains such as `http://dashy.tailscale_device_name/`. I have a very basic understanding of networking but it is definitely my weakest area, although I have read this whole page and I think people are looking for something similar to this. Please correct me if I am wrong, as I would really like to understand how this all works and if it is possible with the current technology.

**Comment by aviadkray** [comment]
my use case:
I have an EKS cluster and I have created a TS_router pod in it. I want to make it so the cluster's API is only accessible from the internal network via the TS_router. I create the EKSs with unique CIDRs, but the "server" address in the kube config, are all under the same domain. For example:
cluster1: https://C991AD720C997160FC5F384158692152.gr7.ap-south-1.eks.amazonaws.com
cluster2: https://3896D0F8BB8331A0EDFE6618FCB19304.gr7.ap-south-1.eks.amazonaws.com
I would like to be able to set the resolver for the specific hosts in the domain, however I can only set one resolver for the entire domain. This makes no sense when there is more then one cluster in the same region.

Thank you.

**Comment by nemani** [comment]
+1 for this feature request.

**Comment by 9at8** [comment]
As an alternative/workaround, here's what I'm doing:

- VM on azure that I use to obtain certs. (you can use a cheap VM or forward port 80 to your machine to grab certs)
  - In order to perform ACME, you will need to add an A entry to this machine's public IP
- Install pi.hole on a machine that will act as a nameserver
  - Add this device to your tailnet
  - Add this device as your tailnet's nameserver (using the tailscale IP address)
- Use pi.hole to add local DNS records, pointing to whatever tailscale IP address you want to point to
- You can now use custom records!

However, this can get very tedious/error-prone to maintain in the long run, so having Tailscale do the magic would be nice.

My setup:
- VM on azure is fetching certs for `service1.home.9at8.dev`, `service2.home.9at8.dev`, `service3.home.9at8.dev` using ACME.
- On my homeserver, I am grabbing certs from the VM.
- My homeserver is on the tailnet, let's say its IP address is `100.x.y.z`.
- My homeserver also runs pi.hole and all the other services (`service1`, `service2`, `service3`) behind a reverse proxy.
- For my pihole config, I have created CNAME records from `service1.home.9at8.dev`, `service2.home.9at8.dev`, and `service3.home.9at8.dev` pointing to `tailscale-ip.home.9at8.dev`, along with an A record for `tailscale-ip.home.9at8.dev` pointing to `100.x.y.z`.
- On the admin console in tailscale, under the DNS tab, I added a nameserver with the IP address `100.x.y.z` (pihole) and split DNS with `home.9at8.dev`.
  - In other words, `*.home.9at8.dev` and `home.9at8.dev` will be resolved by my homeserver when on the tailnet. If not on tailnet, then it will be resolved by the domain name provider (pointing to the azure VM that does ACME and gets certs)

**Comment by jzila** [comment]
I wrote a detailed use case in the above duplicate, but this is a potentially valuable feature for the "progressive adoption" strategy that Tailscale recommends, where you start with a subnet router and then progressively install clients on more endpoints behind the subnet router. Since the Split DNS feature is a way to empower subnet routers, and it prevents resolving to tailnet IPs, a custom DNS solution would give a potential override hiearchy for such a progressive adoption approach.

**Comment by distantorigin** [comment]
My use case:
I would like to be able to alias certain machines with multiple names, so Caddy/Nginx can reverse proxy to them. For example, https://bitwarden, https://minio, etc, would all map back to the same Tailscale IP while on the tailnet. This functionality became of interest to me when trying to do a three-way split DNS--bitwarden.mydomain.com points to a public IP when not on my LAN. While connected to my LAN it points to the internal address at 192.168.xxx.xxx, and when connected to Tailscale it points to the Tailscale IP (otherwise users not on my LAN wouldn't be able to access it--I'm not interested in advertising the LAN IPs as routes).

I run Pi-Hole on the LAN for DNS. To address not being able to have custom records in MagicDNS, I'm currently using a Pi-Hole image in docker that is paired with the Tailscale docker image to host a TS-only Pi-Hole instance as a machine in my Tailnet. One might argue that I could easily use Dnsmasq and forward onto the LAN-based Pi-Hole as upstream, but that's beyond the scope of this comment.

I would love to see custom records for MagicDNS for this reason--it would remove overhead and a lot of friction from my personal infrastructure.

**Comment by 0PandaDEV** [comment]
My use case would be that I have multiple services running on a server that is in my Tailscale network, and then I can make aliases to them that are only internally accessible like for example the tailscale IP of my server is 100.86.67.57, and I have a website running on port 3001 now it this scenario I want it to be accessible through h.lan from any device in my tailnet, so I would need an SRV record and an A record or a mixture of them like with Cloudflare zero trust tunnels.

**Comment by watsonbox** [comment]
This recent blog post might be of interest for those wanting to add specific docker services directly to a tailnet (complete with their own MagicDNS names):

https://tailscale.com/blog/docker-tailscale-guide

**Comment by luis7722** [comment]
My case is simple, I want to use the custom DNS settings to configure my SMTP server to use the MagicDNS domain.

**Comment by matheuscscp** [comment]
+1, this feature would be amazing!

**Comment by bdbusch** [comment]
From our perspective, we have apps running on these servers that have a certain domain for SSL (e.g., abc.com) - and we know that the cert matching/validation logic is based on the host name used to hit the service, so if we can call our machine **umpty** and get an umpty.abc.com alias for the IP, then when the user hits that machine _using the name_, we get a good SSL match.

https://umpty.abc.com:port will find the machine _and_ match the x509. 

Sorry if this use case was mentioned already

**Comment by matheuscscp** [comment]
my use case is:

my friend and I are developing an open source project and we have local k8s clusters running inside our machines via kind. we have manually set some A DNS records inside /etc/hosts in our machines to access each other's clusters through my personal tailnet using domains. i'd like to have a private DNS zone inside my tailnet to automatically set those records in our machines (and keep them in sync as a source of truth)

**Comment by klevis-a** [comment]
Use case: Easy access to SSL-enabled Synology DiskStation without opening ports or configuring a subnet router.

1. If you use Synology's DDNS service you also get an SSL certificate. You are stuck with the my-name.synology.me domain, but the upside is that 1) it's a few clicks to configure 2) you get an SSL certificate.
2. Use the DiskStation's DNS server to configure my-name.synology.me to point to the DiskStation's local IP address and forward other requests to your main DNS server (likely your router/modem).
3. Configure your local devices (laptop, phone, etc.) so their primary DNS server is Synology and backup is router/modem.
4. Now you should be able to simply visit my-name.synology.me and use DiskStation over SSL while pointing to a local IP address.
5. You can easily do the same while using your tailnet as long as you can point my-name.synology.me to your Synology DiskStation node. This removes the need to setup some device (e.g. Synology DiskStation) as a subnet router.

**Comment by rathmerdominik** [comment]
Push from my side as well

This would be an amazing feature, because i would much rather use my own domain than the tailscale one

**Comment by Zizo4418** [comment]
DNS

**Comment by prostolyubo** [comment]
I would love this. My use case is simple: I only want all subdomains of `*.mymachine` to point to `mymachine`.
I have a lot of services running in docker on my server and I'm running into port hell. So I  have a reverse proxy and a local DNS server so I can get rid of docker port mapping and access my services simply by: `nodered.mymachine`, `homepage.mymachine`  instead of `mymachine:1880`, and so on. It works well until I connect from remote location via tailscale. It all falls apart because for example `nodered.mymachine` cannot be resolved.

**Comment by omnizach** [comment]
+1. Reverse proxy on Unraid, I would love to use tailnet URLs and completely get rid of public DNS records with custom domain. Tailscale is awesome and so close to completely solving this problem, missing this one feature. Thanks!

**Comment by Geekdude** [comment]
+1. I would like Tailscale MagicDNS to enable something like having `servicename.mymachine.tailnet-name.ts.net` point to `mymachine.tailnet-name.ts.net:serviceport`. That way, I only have to remember the name of the service without having to remember port numbers.

**Comment by mattdale77** [comment]
+1 would like to run a reverse proxy with service names instead of using port numbers. Some services also insist on being behind https and is the missing link which would allow that natively in tailscale

**Comment by jhsul** [comment]
For anyone who is looking for a solution right now, here is a docker compose stack which I'm using to run an adguard home DNS instance in my tailnet.

I run this stack on the same physical device as my main server (a different docker compose stack), which is also running a tailscale container in sidecar mode - despite being on the same hardware, these appear as two distinct devices with their own IP address in the tailnet. I give my dns device the IP address `100.100.10.10`, and I have that IP configured as the default nameserver.

```yaml
services:
  adguardhome:
    container_name: adguardhome
    image: adguard/adguardhome
    network_mode: service:tailscale-dns
    restart: unless-stopped
    volumes:
      - ../../data/dns/adguardhome/conf:/opt/adguardhome/conf
      - ../../data/dns/adguardhome/work:/opt/adguardhome/work
    profiles:
      - dns
    depends_on:
      - tailscale-dns
  tailscale-dns:
    container_name: tailscale-dns
    image: tailscale/tailscale:stable
    volumes:
      - ../../data/dns/tailscale:/var/lib
      - /dev/net/tun:/dev/net/tun
    cap_add:
      - net_admin
      - sys_module
    env_file:
      - .env
      - .env.local
    command: tailscaled
    privileged: true
    restart: unless-stopped
    profiles:
      - dns
    networks:
      - dns
networks:
  dns:
    name: dns
```
In the .env file, I only have these variables set:
```
TS_ACCEPT_DNS=true
TS_AUTH_KEY=<your auth key>
TS_AUTH_ONCE=false
TS_STATE_DIR=/var/lib/tailscale
TS_USERSPACE=false
```

**Comment by snomiao** [comment]
# Caddy proxy approach

My currently solution for those DNS is setup a local proxy that forward **.fbi.com into correct machine and localservices.

And access local network services by `https://<service-name>.<machine-name>.fbi.com`

## Example

My Caddyfile for reference: (same config for server)

```caddyfile
{
	servers {
		trusted_proxies static 0.0.0.0/0
	}
}

# route fbi internal
https://127.0.0.1, https://localhost {
	tls internal
	redir https://fbi.com
}
https://fbi.com {
	tls internal
	reverse_proxy http://fbi.com {
		header_up Host fbi.com
	}
}
# forward HTTPS -> HTTP
https://*.fbi.com {
	tls internal
	@proxyhostport header_regexp subdomains Host (.+?).fbi.com
	reverse_proxy @proxyhostport http://fbi.com {
		header_up Host {re.subdomains.1}.fbi.com
	}
}
http://*.fbi.com {
	respond 502
}

# ====================================================
# hosts

# route to hosts
https://*.host1.fbi.com {
	tls internal
	@proxyhostport header_regexp service Host (.+?).host1.fbi.com
	reverse_proxy @proxyhostport http://host1 {
		header_up Host {re.service.1}.fbi.com
	}
}
https://*.host2.fbi.com {
	tls internal
	@proxyhostport header_regexp service Host (.+?).host2.fbi.com
	reverse_proxy @proxyhostport http://host2 {
		header_up Host {re.service.1}.fbi.com
	}
}
https://*.host3.fbi.com {
	tls internal
	@proxyhostport header_regexp service Host (.+?).host3.fbi.com
	reverse_proxy @proxyhostport http://host3 {
		header_up Host {re.service.1}.fbi.com
	}
}

# ==================================================
# services
# route to services
http://vscode.fbi.com {
	# hack: fix xfh for code
	@xfh header_regexp xfh X-Forwarded-Host (.*)
	reverse_proxy @xfh :8000 {
		header_up X-Forwarded-Host {re.xfh.1}
	}
	reverse_proxy :8000
}

# activitywatch
http://activitywatch.fbi.com {
	reverse_proxy :5600
}
# alias
http://aw.fbi.com {
	reverse_proxy :5600 {
		header_up Host localhost
	}
}

# everything
http://everything.fbi.com {
	reverse_proxy :2489
}
# alias
http://et.fbi.com {
	reverse_proxy :2489
}

```

for example setup

and then you can access aw service by


activitywatch on localhost
https://aw.fbi.com/#/home

activitywatch on host1
https://aw.host1.fbi.com/#/home

activitywatch on host2
https://aw.host2.fbi.com/#/home

![image](https://github.com/user-attachments/assets/dd483e45-05c5-4219-b5a3-72aa78252424)

## extend

Btw you could also expose this network into public by forwarding into `*.<your-domain>` to `*.fbi.com`

Hopefully someone could enjoy this idea.

## references

- [Caddy - The Ultimate Server with Automatic HTTPS]( https://caddyserver.com/ )
- [web - Why is fbi.com localhost? - Stack Overflow]( https://stackoverflow.com/questions/65274602/why-is-fbi-com-localhost )
- [Hosting a project locally? You can access it via http://FBI.com : r/webdev]( https://www.reddit.com/r/webdev/comments/1cc90m3/hosting_a_project_locally_you_can_access_it_via/ )

**Comment by prostolyubo** [comment]
Guys, those are really nice WORKAROUNDS, but the idea here is to request Tailscale to implement a solution that would not require from us to make such a workarounds and greatly simplify the process.

**Comment by sbrocket** [comment]
> # Caddy proxy approach
> My currently solution for those DNS is setup a local proxy that forward **.fbi.com into correct machine and localservices.
> 
> And access local network services by `https://<service-name>.<machine-name>.fbi.com`
> 
> ## Example
> 1. install caddy server https://caddyserver.com/
> 2. setup Caddyfile and run it when system start
> 
> a example Caddyfile for reference: (same config for every hosts)
> 
> ```Caddyfile
> {
> 	servers {
> 		trusted_proxies static 0.0.0.0/0
> 	}
> }
> 
> # route fbi internal
> https://127.0.0.1, https://localhost {
> 	tls internal
> 	redir https://fbi.com
> }
> https://fbi.com {
> 	tls internal
> 	reverse_proxy http://fbi.com {
> 		header_up Host fbi.com
> 	}
> }
> # forward HTTPS -> HTTP
> https://*.fbi.com {
> 	tls internal
> 	@proxyhostport header_regexp subdomains Host (.+?).fbi.com
> 	reverse_proxy @proxyhostport http://fbi.com {
> 		header_up Host {re.subdomains.1}.fbi.com
> 	}
> }
> http://*.fbi.com {
> 	respond 502
> }
> 
> # ====================================================
> # hosts
> 
> # route to hosts
> https://*.host1.fbi.com {
> 	tls internal
> 	@proxyhostport header_regexp service Host (.+?).host1.fbi.com
> 	reverse_proxy @proxyhostport http://host1 {
> 		header_up Host {re.service.1}.fbi.com
> 	}
> }
> https://*.host2.fbi.com {
> 	tls internal
> 	@proxyhostport header_regexp service Host (.+?).host2.fbi.com
> 	reverse_proxy @proxyhostport http://host2 {
> 		header_up Host {re.service.1}.fbi.com
> 	}
> }
> https://*.host3.fbi.com {
> 	tls internal
> 	@proxyhostport header_regexp service Host (.+?).host3.fbi.com
> 	reverse_proxy @proxyhostport http://host3 {
> 		header_up Host {re.service.1}.fbi.com
> 	}
> }
> 
> # ==================================================
> # services
> # route to services
> http://vscode.fbi.com {
> 	# hack: fix xfh for code
> 	@xfh header_regexp xfh X-Forwarded-Host (.*)
> 	reverse_proxy @xfh :8000 {
> 		header_up X-Forwarded-Host {re.xfh.1}
> 	}
> 	reverse_proxy :8000
> }
> 
> # activitywatch
> http://activitywatch.fbi.com {
> 	reverse_proxy :5600
> }
> # alias
> http://aw.fbi.com {
> 	reverse_proxy :5600 {
> 		header_up Host localhost
> 	}
> }
> 
> # everything
> http://everything.fbi.com {
> 	reverse_proxy :2489
> }
> # alias
> http://et.fbi.com {
> 	reverse_proxy :2489
> }
> ```
> 
> 3. and then for example, you can access aw service by
> 
> activitywatch on localhost https://aw.fbi.com/#/home
> 
> activitywatch on host1 https://aw.host1.fbi.com/#/home
> 
> activitywatch on host2 https://aw.host2.fbi.com/#/home
> 
> https://aw.jr.fbi.com	https://vscode.fbi.com
> ![image](https://github.com/user-attachments/assets/dd483e45-05c5-4219-b5a3-72aa78252424)	![image](https://github.com/user-attachments/assets/982a6cf3-1660-4a2e-9722-914857b0d346)
> ## extend
> Btw you could also expose this network into public by forwarding into `*.<your-domain>` to `*.fbi.com`
> 
> Hopefully someone could enjoy this idea.
> 
> ## references
> * [Caddy - The Ultimate Server with Automatic HTTPS](https://caddyserver.com/)
> * [web - Why is fbi.com localhost? - Stack Overflow](https://stackoverflow.com/questions/65274602/why-is-fbi-com-localhost)
> * [Hosting a project locally? You can access it via http://FBI.com : r/webdev](https://www.reddit.com/r/webdev/comments/1cc90m3/hosting_a_project_locally_you_can_access_it_via/)

The fact that whoever owns fbi.com directed it at 127.0.0.1 is funny, but it seems like a *really* bad idea to rely on a domain you don’t control and which you can’t issue anything more than locally-trusted certificates for. That will not give you HTTPS worth the name and is not a setup I’d recommend generally.

Tailscale team: If you needed more justification for why this feature is important than the hundreds of votes here, this should be it. This is complex enough that people are going to get it wrong in subtle but important ways and end up with insecure setups.

**Comment by sevenrats** [comment]
is there any fundamental blocker to this?  i noticed the TODO here:
https://github.com/tailscale/tailscale/blob/6edf357b96b28ee1be659a70232c0135b2ffedfd/ipn/ipnlocal/local.go#L2994

My two cents is this:
I would implement cnames probably if I had any indicator that it would be accepted.
It would also be cool to automatically return records for machines with a specific tag.  like if I lookup a particular record, x.domain.tld, then an A record would be returned for every machine with the tag tag:dns:x
or even something super cool like tag:dns:roundrobin:x versus tag:dns:failover:x

**Comment by almereyda** [comment]
The use case for resolving tags is tracked in https://github.com/tailscale/tailscale/issues/4324.

**Comment by dolceAlka** [comment]
This would be great short even for workarounds. Because tailscale uses ULA addressing for IPv6, validating resolvers will not resolve domains because of potential dns rebinding attacks. Short of migrating the entire tailscale addressing scheme, it would be ideal to have this.

**Comment by jucor** [comment]
For anyone needing these custom record, wouldn't [https://almeidapaulopt.github.io/tsdproxy/docs/](TSDProxy) solve the issue?  It was advertised in Tailscale November newsletter. @almeidapolaut , what do you think ?

**Comment by madscientist16** [comment]
Your link points to the wrong page https://almeidapaulopt.github.io/tsdproxy/

Also from what I understand tsdproxy only works for containers. It does help somewhat since I won't need to create a seperate tailscale container for each service.

**Comment by almeidapaulopt** [comment]
> Your link points to the wrong page https://almeidapaulopt.github.io/tsdproxy/
> 
> Also from what I understand tsdproxy only works for containers. It does help somewhat since I won't need to create a seperate tailscale container for each service.

this will change soon. you'll be able to add services that aren't running in docker

**Comment by marjacob** [comment]
I need custom SRV and TXT records for my use case. As far as I can see tsdproxy won't be able to help with that?

**Comment by jucor** [comment]
@madscientist16 thanks for the flag on the link, fixed!

**Comment by dolceAlka** [comment]
> For anyone needing these custom record, wouldn't [TSDProxy](https://almeidapaulopt.github.io/tsdproxy/docs/) solve the issue? It was advertised in Tailscale November newsletter. @almeidapaulopt , what do you think ?

It would be much better to handle this outside of a specific niche project, and allow for more flexible domaining and analytics

**Comment by jucor** [comment]
Sure. I'm mostly flagging a community-based workaround until Tailscale
supports it :) A few useful aspects of tailscale started that way -- for
example the Synology packages were a third party work originally, until
Tailscale had the bandwidth to handle it and take it over.

**Comment by tianon** [comment]
I've noted it in https://github.com/tailscale/tailscale/issues/4457#issuecomment-2578767140, but I'll reiterate here -- reading the history, it sounds like this is primarily blocked on it being risky to allow unmitigated custom DNS records, but the ask over in #4457 was just for the more targeted "add an alias to an existing hostname" which would be immensely helpful and seems a lot less risky to support than the wider custom access proposed here.  (I'd also guess it would meet 80% of the use cases.)

**Comment by phiberoptick** [comment]
Just my 2c here.

I agree that this should above all not compromise security and secondarily the implementation should be as transparent and non-invasive as possible.  We don't want a feature to end up in a future CVE.

I'm not sure about other's configurations, but in mine, the internal DNS servers in question also happen to be on my tailnet.  Either via a subnet router or as a service inside of a docker stack.

I remember that Alex Kretzschmar mentioned using split dns in a podcast and/or blog to handle a similar case.  Could that possibly be useful in some of these situations?  

Please ignore me if this has been mentioned already.  I believe I saw a link to one of his articles above from @watsonbox as well.

Here's the link I am referring to:
[Link](https://blog.ktz.me/splitdns-magic-with-tailscale/)

Apparently he also made a video.  🤣 
[Link](https://youtu.be/Uzcs97XcxiE)

**Comment by jucor** [comment]
Yes, splitDNS is my current setuo, running my own tailnet-only DNS server for a custom domain as a crutch. But it's far from as smooth as having CNAME entries in Tailscale DNS: 
- it requires a DNS server, active at all times
- and, unless I'm mistaken, doesn't allow to manage "something-something.ts.net"
- which also stands in the way of the awesome automatic tailscale certs handling.

All of the above would be magically solved by allowing CNAME entries (aka aliases) in Tailscale DNS.

**Comment by jucor** [comment]
> I'm not sure about other's configurations, but in mine, the internal DNS servers in question also happen to be on my tailnet.  Either via a subnet router or as a service inside of a docker stack.
> 
> I remember that Alex Kretzschmar mentioned using split dns in a podcast and/or blog to handle a similar case.  Could that possibly be useful in some of these situations?  

Yes, splitDNS is my current setup, running my own tailnet-only DNS server for a custom domain as a crutch. But it's far from as smooth as having CNAME entries in Tailscale DNS: 
- it requires a DNS server, active at all times
- and, unless I'm mistaken, doesn't allow to manage "something-something.ts.net"
- which also stands in the way of the awesome automatic tailscale certs handling.

All of the above would be magically solved by allowing CNAME entries (aka aliases) in Tailscale DNS.

**Comment by prostolyubo** [comment]
> For anyone needing these custom record, wouldn't [TSDProxy](https://almeidapaulopt.github.io/tsdproxy/docs/) solve the issue? It was advertised in Tailscale November newsletter. @almeidapaulopt , what do you think ?

Finding a WORKAROUND is not the same as SOLVING the issue. Especially that a workaround in this case is completely missing the point. There are plenty of workarounds to employ but first and foremost we want a Quality of Life improvement to not have to mess around with other systems/third party software/hardware/etc.

**Comment by btrepp** [comment]
Also with running your own DNS servers for anyone exploring this as an option. DNS servers are a pain.

There's no easy way for HA, and thus if whatever server is running your DNS workaround is having issues, you may get weird errors and problems depending on how you configure it.

As tailscale is kinda already doing DNS for you it is the natural place for the feature to live.

Potentially for security concerns it could be configured/allowed in the policy JSON files?.

Without this, it's a bit tricky to offer highly available services in a tailnet.

**Comment by tpanum** [comment]
> Also with running your own DNS servers for anyone exploring this as an option. DNS servers are a pain.
> 
> There's no easy way for HA, and thus if whatever server is running your DNS workaround is having issues, you may get weird errors and problems depending on how you configure it.
> 
> As tailscale is kinda already doing DNS for you it is the natural place for the feature to live.
> 
> Potentially for security concerns it could be configured/allowed in the policy JSON files?.
> 
> Without this, it's a bit tricky to offer highly available services in a tailnet.

This answer is truly the essence of it all.

**Comment by wsw70** [comment]
Not having the ability to add "CNAME" records to a Tailsclae, MagicDNS-provided name (`server.hello-world.ts.net`) makes it very difficult to server multiple web services on that server (usually served by an edge router/reverse proxy).
They require (very roughly speaking) that you address them via a name that will redirect the call to the specific service. In other words, when I have service `wazaa` running on the server `server.hello-world.ts.net` I need to use `https://wazaa.hello-world.ts.net`.

From what I understand (and saw in the POC I did) it is today impossible to say "`wazaa.hello-world.ts.net` has the same IP as `wazaa.hello-world.ts.net`" (aka A record in DNS) or "add the alias `wazaa.hello-world.ts.net` to point to `server.hello-world.ts.net`" (aka CNAME in DNS).

This is really annoying because as others said, you need to run **and maintain the zone** of another, separate DNS. It is possible to have wildcard DNS entries that may ease the pain but it would bork, at best, in a home deployment because it has zero scalability. This of course requires Tailscale to delegate `hello-world.ts.net` which I do not believe can be done (practically, not technically).

In other words, web services on a Tailscale network are not easy, if at all possible

**Comment by benedictjohannes** [comment]
> They require (very roughly speaking) that you address them via a name that will redirect the call to the specific service. In other words, when I have service `wazaa` running on the server `server.hello-world.ts.net` I need to use `https://wazaa.hello-world.ts.net`.

I have a much simpler suggestion.

have magicDNS publish `*.server.hello-world.ts.net` as the same IP address with `server.hello-world.ts.net`

so each tailnets still handle DNS resolution for each Tailscale **device**, but with the addition of a wildcard DNS subdomain for the same device IP.

That make it trivial to have _wazaa.server.hello-world.ts.net_ or _basicallyAnything.server.hello-world.ts.net_ 

This should be quite straightforward to implement compared to promoting service into tailnet subdomain. Wdyt?

I believe this provides a way to sidestep this issue entirely, but doesn't address the problem directly. Should I open a separate issue instead?

**Comment by qm3ster** [comment]
`wazaa.server.hello-world.ts.net` is useful for home (and maybe all wildcards undermachines should be enabled for DNS and TLS by default) but it doesn't have the flexibility of moving the service to another machine unlike CNAMEs.

**Comment by spkis** [comment]
Well, for a temporary fix "have magicDNS publish *.server.hello-world.ts.net as the same IP address with server.hello-world.ts.net" would still have more flexibility as the current implementation!

**Comment by chingiz-ace1m** [comment]
It would be a great feature to have. Thanks Tailscale

**Comment by peskyAdmin** [comment]
> Well, for a temporary fix "have magicDNS publish *.server.hello-world.ts.net as the same IP address with server.hello-world.ts.net" would still have more flexibility as the current implementation!

Came here to basically say the same thing. It would be great to support Wildcard subdomains for the TS full domain. Ran into this issue when trying to deploy Socioboard via tailscale docker sidecar. That project architecture has a built in proxy that handles routes through subdomains.

**Comment by babinskiy** [comment]
Why not have the option to define DNS names similarly to ACLs? For example:

```
dns: { "some-service": ["server1", "server2"], "other-service": ["server3"] }
```

These could be populated into the internal DNS service as `A` records like some-service.tailXXX.ts.net and other-service.tailXXX.ts.net, each resolving to multiple IPs..

This feature doesn't seem too difficult to implement but would be really useful.

**Comment by movchan74** [comment]
>  Well, for a temporary fix "have magicDNS publish *.server.hello-world.ts.net as the same IP address with server.hello-world.ts.net" would still have more flexibility as the current implementation!

Wildcard support would be extremely useful. I’m using [Coolify](https://coolify.io/) for my self-hosted services. It uses Traefik to route subdomains to specific applications. Wildcard support would allow me to assign a subdomain for each application for easy access.

**Comment by prostolyubo** [comment]
> These could be populated into the internal DNS service as `A` records like some-service.tailXXX.ts.net and other-service.tailXXX.ts.net, each resolving to multiple IPs..
> 
> This feature doesn't seem too difficult to implement but would be really useful.

But seems that it might use additional resources on their side. While the **wildcard idea is simple, cheap and even easier to implement and maintain.** Then you can just have a reverse proxy on your tailnet and handle whatever you want yourself.

**Comment by spkis** [comment]
Hi,

They just introduced this 4via6 thing for subnet routers etc.
Would it not be possible to hop on that for services as well or use similar logic?!



Ricardo Kissinger
Head of IT Infrastructure and IT Security

IT Infra+Sec
[juniz_Energy_balanced_blue.jpg]
[cid:C2_signature_linkedin_32x32_927d87ef-23c2-4fb3-9794-500d902fa4bf.png]<https://www.linkedin.com/company/juniz/>

ju:niz Energy GmbH
Einsteinring 43 | 85609 Aschheim

M: +49 (151) 150 46 204
E-Mail: ***@***.***
Web: www.juniz.com
Sie können hier ***@***.***/> auch direkt einen Termin buchen oder mit mir in Teams ***@***.***> chatten.


ju:niz Energy GmbH
Sitz der Gesellschaft: Aschheim
AG München, HRB 277043
USt-ID: DE 299 706 389
Geschäftsführung: Dr. Marcus Groll | Markus Binding | Gabriele Schmiedel



From: Heti ***@***.***>
Date: Saturday, 24. May 2025 at 14:45
To: tailscale/tailscale ***@***.***>
Cc: spkis ***@***.***>, Comment ***@***.***>
Subject: Re: [tailscale/tailscale] FR: Support custom records in MagicDNS (#1543)

[https://avatars.githubusercontent.com/u/45892416?s=20&v=4]prostolyubo left a comment (tailscale/tailscale#1543)<https://github.com/tailscale/tailscale/issues/1543#issuecomment-2906813624>

These could be populated into the internal DNS service as A records like some-service.tailXXX.ts.net and other-service.tailXXX.ts.net, each resolving to multiple IPs..

This feature doesn't seem too difficult to implement but would be really useful.

But seems that it might use additional resources on their side. While the wildcard idea is simple, cheap and even easier to implement and maintain. Then you can just have a reverse proxy on your tailnet and handle whatever you want yourself.

**Comment by Misza13** [comment]
Came here as another homelab owner to say that even the simple wildcard solution of pointing all of `*.machine.tailnet.ts.net` to be equivalent to `machine.tailnet.ts.net` would solve a lot of issues as I could sort out the rest in my Caddy configuration.

As many pointed out, many services do not properly support being hosted in "subfolders" and this CANNOT be trivially fixed with rewrites at the proxy level. Also, there is the browser security issue with that.

Currently, I am forced to host different services at different ports, making me feel like I'm in the 90s again. And it introduces different problems. For example, BitWarden does not seem to be able to distinguish services by port, so on every service I host, it displays a list of all my credentials (which, by now, is a very long list indeed).

**Comment by infogulch** [comment]
Have you looked into the newish app connectors feature? I glanced through the docs briefly and it looks like it might work for this use case. https://tailscale.com/kb/1281/app-connectors

**Comment by Misza13** [comment]
> Have you looked into the newish app connectors feature? I glanced through the docs briefly and it looks like it might work for this use case. https://tailscale.com/kb/1281/app-connectors

From the docs:

> Configure a Linux device on your tailnet to act as the designated app connector that will route the traffic to the application. This device must have a public IP address, and IP forwarding must be enabled.

This might actually be a no-go for many users, including myself. One reason might be the lack of a static IPv4 address. Another might be that some (most?) homelab setups do not actually expose any services to the Internet - only to client devices belonging to the same tailnet.

The whole point of using Tailscale, at least for me, was to have all my servers and clients in a network that feels like a LAN, but without the hassle of having to manage my own DNS and routes.

**Comment by ndlarsen** [comment]
> The whole point of using Tailscale, at least for me, was to have all my servers and clients in a network that feels like a LAN, but without the hassle of having to manage my own DNS and routes.

This + the lack of need for publicly exposed ports....

**Comment by spkis** [comment]
For us, having a public exposed address also does not work, since we simply don’t have one on most areas of the business.



Ricardo Kissinger
Head of IT Infrastructure and IT Security

[juniz_Energy_balanced_blue.jpg]
[cid:C2_signature_linkedin_32x32_927d87ef-23c2-4fb3-9794-500d902fa4bf.png]<https://www.linkedin.com/company/juniz/>

ju:niz Energy GmbH
Einsteinring 43 | 85609 Aschheim

M: +49 (151) 150 46 204
E-Mail: ***@***.***
Web: www.juniz.com
Sie können hier ***@***.***/> auch direkt einen Termin buchen oder mit mir in Teams ***@***.***> chatten.


ju:niz Energy GmbH
Sitz der Gesellschaft: Aschheim
AG München, HRB 277043
USt-ID: DE 299 706 389
Geschäftsführung: Dr. Marcus Groll | Markus Binding | Gabriele Schmiedel



From: Nicolas Damgaard Larsen ***@***.***>
Date: Friday, 4. July 2025 at 19:39
To: tailscale/tailscale ***@***.***>
Cc: spkis ***@***.***>, Comment ***@***.***>
Subject: Re: [tailscale/tailscale] FR: Support custom records in MagicDNS (#1543)

[https://avatars.githubusercontent.com/u/356953?s=20&v=4]ndlarsen left a comment (tailscale/tailscale#1543)<https://github.com/tailscale/tailscale/issues/1543#issuecomment-3036946695>

The whole point of using Tailscale, at least for me, was to have all my servers and clients in a network that feels like a LAN, but without the hassle of having to manage my own DNS and routes.

This + the lack of need for publicly exposed ports....

**Comment by carsaig** [comment]
+1 add custom domain and wildcard support as a feature, and - super important: make the funnel feature expose custom ports! Those three features tied together would make this an unbeatable simple option for homelab and small to midsize firms!
The funnel feature is explicitly important because it allows to expose singular services to the outside world while keeping everything else behind closed doors. Super convenient. No hassle with hardening all along the levels.
I know - routing, reverse proxying, DNS management - all nasty stuff and not simple to implement. Surely painful. But your 90% there - just stack this on top to make it slick as hell.

**Comment by sebastianrasor** [comment]
~~The current name of this FR is "FR: Support custom records in MagicDNS"~~

~~Tailscale (and Headscale) now support extra records of types A and AAAA. I feel like at this point we need to either close this FR as completed, or rename this feature request to better reflect what we've all been asking for: CNAME support (aliases) and wildcard records.~~

Tailscale doesn't actually support extra records currently.

Just to add my 2 cents to this, I'd really like to be able to just set a CNAME record as an extra record in my Headscale config which points to a MagicDNS name so I can stop trying to juggle public DNS records pointing to "static" tailnet IP addresses or trying to figure out a proper local hosted split DNS. Both of these solutions rely on the assumption that a node in a tailnet is static and will never change IP addresses, which for the most part is a safe assumption to make until you want to wipe a node and start over, then it gets a new IP and you have to track down everywhere you've hard coded the previous IP. If Tailscale supported CNAME extra records, all I'd have to do is make sure to delete the old node, make sure the new node name matches, and all is good.

**Comment by netthier** [comment]
> Tailscale (and Headscale) now support extra records of types A and AAAA. I feel like at this point we need to either close this FR as completed, or rename this feature request to better reflect what we've all been asking for: CNAME support (aliases) and wildcard records.

This is news to me, could you elaborate on what feature you're referencing? Is it actually possible to create custom A records and e.g. have multiple tailnet domains point at the same tailnet node?

**Comment by sebastianrasor** [comment]
> This is news to me, could you elaborate on what feature you're referencing? Is it actually possible to create custom A records and e.g. have multiple tailnet domains point at the same tailnet node?

https://headscale.net/stable/ref/dns/#setting-extra-dns-records

I don't know what the alternative is for tailscale.com since I've only ever used Headscale.

Here's how I use extra records:
https://github.com/sebastianrasor/nix-config/commit/e2acbda13d3e730a1c3fd5d7e39b4505a1820d85

Instead of including 6 copies of a hard coded IP address, I'd like to be able to just CNAME them to the existing MagicDNS record for my server.

**Comment by netthier** [comment]
I'm not aware of an equivalent feature in regular Tailscale, but I'd agree that it would help substantially with most of the use cases described.
But without a confirmation that this feature is also available in Tailscale (I couldn't find it) this issue should not be closed.

It is possible to set up custom nameservers in MagicDNS, but that's substantial effort for most users and AFAIK does not support creating custom records for tailnet domains.

**Comment by madscientist16** [comment]
Extra DNS records is a headscale only feature. It says so in their docs.

https://headscale.net/stable/about/features/

**Comment by sebastianrasor** [comment]
That's my mistake, I got a bit overzealous assuming if it exists in Headscale it should exist in Tailscale lol

**Comment by carsaig** [comment]
> I'm not aware of an equivalent feature in regular Tailscale, but I'd agree that it would help substantially with most of the use cases described. But without a confirmation that this feature is also available in Tailscale (I couldn't find it) this issue should not be closed.
> 
> It is possible to set up custom nameservers in MagicDNS, but that's substantial effort for most users and AFAIK does not support creating custom records for tailnet domains.

correct. Setting up custom nameservers doesn't go as far as required for full custom domains. It just does some rewrite logic for TS domains. It's not just DNS name handling, it's certificate management, routing, headers, reverse-proxying etc. and that's the missing piece in the Tailscale setup. At least when it comes to custom (non Tailscale) domains. They do it for their tailnet domains and that works fine (as long as no special ports are required^^).
Split DNS is a mess. I have it in place and I have custom nameservers in place but that's not enough to tackle the goal.

Correct me, if I'm wrong - but the Tailscale feature sounds like the [tailscale serve option](https://tailscale.com/blog/reintroducing-serve-funnel), that allows serving any internal application on a custom tailnet domain without adding ports to the URL. Though it is limited to Port 8443, 443 and 1000 I think. Thus not very flexible. The Headscale feature looks as if it replicated the Tailscale feature tom some extent, exposing the reverse-proxy settings which gives a little bit of extra config head-room - but I'm not sure at first glance if that is enough to handle full custom domains - although, headscale runs on owned custom domains so it might be a solution. No idea, I never configured Headscale...

**Comment by Misza13** [comment]
> Correct me, if I'm wrong - but the Tailscale feature sounds like the [tailscale serve option](https://tailscale.com/blog/reintroducing-serve-funnel), that allows serving any internal application on a custom tailnet domain without adding ports to the URL.

`tailscale serve` doesn't do anything that any commonly used reverse proxy (Caddy, nginx) can't already do. In fact, it's limited to proxying only *local* services.

A common (I think) homelab architecture is a single server with a proxy and Tailscale running. The proxy has multiple upstreams defined and could use vhosts to direct local traffic (`vaultwarden.mylab`, `grafana.mylab` etc.) to local services, containers, or even other machines on the network. But since on the Tailnet it only can have one hostname (say, `services.foo-bar.ts.net`), it needs to do it by ports.

To go around it, you'd have to have a *separate* instance of Tailscale running for *each* service (`vaultwarden.foo-bar.ts.net`, `grafana.foo-bar.ts.net` etc.), implying a separate instance of the proxy too (if you want SSL for services that don't provide it, or don't want their self-signed certificates, and don't feel like running your CA, it's a rabbit hole - remember, Tailscale works with certbot so, for me, Caddy auto-provisions certificates). That implies some container shenanigans as now you need to bridge the three (Tailscale + proxy + service) together (nstead of the host). And you can run into device limit on a free plan (thank you, Tailscale, for providing a free plan, btw!)

Yeah, it's a mess and I guess solutions out there exist, but for me, that would require switching to Headscale (which would require me to run my own coordination server; again, thank you, Tailscale, for providing that for free!), running a full-blown DNS (both of which become my personal point of failure) and learn how routing actually works.

**Comment by TravisH84** [comment]
Would also love to see support for custom DNS, seems like such a waste to have to redirect a specific domain to a nameserver, would be ideal for a limited number to be able to just support split DNS through a custom DNS implementation in tailscale itself. Would be great to be able to point a custom DNS entry to a tailscale IP and not have to worry about having a DNS server somewhere do the rewrites and so on, just do it all in tailscale given its already doing it for Magic DNS and so on.

I know there are cert issues and so on, but why not start with just custom DNS, and then potentially expand from there, so we coudl say X.Y.com points to 100.x.x.x address and be done with it all :).

**Comment by kabirsikand** [comment]
While not exactly supporting custom MagicDNS names, a lot of requests in here can be solved with Tailscale Services. Read more here: https://tailscale.com/blog/services-beta

Keeping open for now as we may still move to support custom MagicDNS names.

**Comment by srid** [comment]
> While not exactly supporting custom MagicDNS names, a lot of requests in here can be solved with Tailscale Services. Read more here: https://tailscale.com/blog/services-beta

I can confirm this to work! But, would it also be possible to expose these services to the public internet?

**Comment by quentinsf** [comment]
As soon as I saw the Services announcement, I thought, 'Aha! Maybe this will effectively provide me virtual host names for web services.'

It looks good, and I would have dived straight into trying it except, I must confess, that I'm cheap, and this is just for personal projects at present, so I'll be waiting to see whether Services continue to be available on the free plan...  🙂

**Comment by tianon** [comment]
The announcement explicitly says they'll grandfather all "legacy usage" even if this becomes a paid feature ☺️

(In my experience, they're also extremely generous with those sorts of things on the whole, as a matter of culture and policy; they've got a few great blog posts on that topic too)

**Comment by jucor** [comment]
> I can confirm this to work! But, would it also be possible to expose these services to the public internet?

The blog post mentions that they have in mind to allow hooking up services with funnels, to expose them to the public internet. That's part of their ideas, but not yet done.

**Comment by tris** [comment]
For anyone who needs to expose services to the internet *now*, I'd probably still recommend [tsdproxy](https://github.com/almeidapaulopt/tsdproxy) (or my own quite dumb & older one, [funnelproxy](https://github.com/tris/funnelproxy)).

Hats off to the Tailscale team for the Services release; it looks cool and I'm excited to try it.

**Comment by mattdale77** [comment]
> For anyone who needs to expose services to the internet _now_, I'd probably still recommend [tsdproxy](https://github.com/almeidapaulopt/tsdproxy) (or my own quite dumb & older one, [funnelproxy](https://github.com/tris/funnelproxy)).
> 
> Hats off to the Tailscale team for the Services release; it looks cool and I'm excited to try it.

Bear in mind that tsdproxy has been abandoned. There are other options such as caddy, Traefik and TSD Bridge I think which can liaise with magic DNS. I'd be keen for the Services functionality to come to Headscale

**Comment by fitz-vivodyne** [comment]
@kabirsikand is there a way to deploy a service instance via tsnet (golang)?

I've been using a simple go tsnet based docker image as a reverse proxy sidecar to make ECS tasks available on our tailnet.
I would like to be able to port that over to use services, but I'm not sure what the correct way of doing that would be.

Edit: Tracking issue @ https://github.com/tailscale/tailscale/issues/17697

**Comment by kabirsikand** [comment]
In hopes of not getting too off-topic from the core request here (custom MagicDNS names) I'll answer a few of the above:
- We're working on tsnet support! This will likely be a set of APIs similar to Listen/ListenTLS, but for Services. So should be a pretty trivial migration for existing setups.
- Funnel support is on our radar, but not actively in progress. May be worth an FR in this repo so we can track demand etc... and personally, I'd like to get that hooked up!

On the topic of pricing, yes. We're going to honor your use today for the purposes of testing, getting feedback, finding issues, understanding your use cases, etc. That's all much more valuable than any minor benefits of usage-based pricing for a beta product! As @tianon points out, we have a strong ethos around not "pulling the rug out" from under you.

**Comment by mdziczkowski** [comment]
Hey, any news about the server-side implementation of this and similiar and / or releated features?

**Comment by ankycooper** [comment]
Allowing custom domains for MagicDNS would be a game-changer for both homelab and enterprise users. Many organizations already have established internal DNS naming standards (e.g., `app.company.com`, `db.company.com`, `grafana.company.com`), and being forced to use the tailnet domain creates inconsistency, breaks existing documentation, complicates migrations, and introduces unnecessary operational overhead. Supporting custom domains would allow users to seamlessly integrate Tailscale into their existing infrastructure, simplify service discovery, improve the user experience, and make Tailscale feel like a natural extension of their network rather than a separate overlay requiring its own naming convention.

This feature has been requested and actively discussed for more than five years, demonstrating sustained demand from the community rather than a niche or short-lived request. As Tailscale continues to evolve from a VPN into a broader networking and infrastructure platform, support for custom DNS domains feels like a natural next step. It would help organizations preserve existing naming standards, reduce friction during adoption, and unlock cleaner integrations with internal services. Given the maturity of the product and the longevity of this request, implementing custom domains for MagicDNS would deliver meaningful value to a large segment of the Tailscale user base.

**Comment by acffordyce973** [comment]
> Allowing custom domains for MagicDNS would be a game-changer for both homelab and enterprise users. Many organizations already have established internal DNS naming standards (e.g., `app.company.com`, `db.company.com`, `grafana.company.com`), and being forced to use the tailnet domain creates inconsistency, breaks existing documentation, complicates migrations, and introduces unnecessary operational overhead. Supporting custom domains would allow users to seamlessly integrate Tailscale into their existing infrastructure, simplify service discovery, improve the user experience, and make Tailscale feel like a natural extension of their network rather than a separate overlay requiring its own naming convention.
> 
> This feature has been requested and actively discussed for more than five years, demonstrating sustained demand from the community rather than a niche or short-lived request. As Tailscale continues to evolve from a VPN into a broader networking and infrastructure platform, support for custom DNS domains feels like a natural next step. It would help organizations preserve existing naming standards, reduce friction during adoption, and unlock cleaner integrations with internal services. Given the maturity of the product and the longevity of this request, implementing custom domains for MagicDNS would deliver meaningful value to a large segment of the Tailscale user base.

Could not have said this anywhere near as good myself. The fact that this has been a request for 5 years already tells me it isn't happening though.

**Comment by mdziczkowski** [comment]
Since this issue was made on my request, I would like to add something from myself ;-)

---

**NOTE**: It was forgotten to become mentioned by this issue creation that this feature should be avaliable as default (build-in) free of any charge, independantly from the plan the user has.

---

It would be great if the MagicDNS would support (as build-in feature) all existing kinds (no only the common ones) of DNS records.

This would make things easier for all those who host services on their machines (or in their infrastructure) and thru this they would't  need to setup additional DNS server

**Comment by MitchTalmadge** [comment]
> this feature should be avaliable as default (build-in) free of any charge

While that would be nice, it's ultimately not our place to declare such requirements.

**Comment by michaelchiche** [comment]
> Since this issue was made on my request, I would like to add something from myself ;-)
> 
> **NOTE**: It was forgotten to become mentioned by this issue creation that this feature should be avaliable as default (build-in) free of any charge, independantly from the plan the user has.
> 
> It would be great if the MagicDNS would support (as build-in feature) all existing kinds (no only the common ones) of DNS records.
> 
> This would make things easier for all those who host services on their machines (or in their infrastructure) and thru this they would't need to setup additional DNS server

it's open source. fork it and push a pr

**Comment by tthoma24** [comment]
> > this feature should be avaliable as default (build-in) free of any charge
> 
> While that would be nice, it's ultimately not our place to declare such requirements.

Why are you bootlicking for a corporation? It is absolutely our place as users of the service to make these requests.

**Comment by mdziczkowski** [comment]
@MitchTalmadge 

> While that would be nice, it's ultimately not our place to declare such requirements.

This wasn't a requirment, but a kind proposal of improvement, with would for sure boost the user experience and give additional promotion, without need for the company to pay high marketing fee's and thereof - thru saving a loot of money - they could redirect their powers to make Tailscale even better and hopefully to make more corporate-level alike features free of charge ;-)

---

Last time I also forgot that about adding the DNS 100.100.100.100 should be discoverable - for example, if I would like to (as temporare solution until I don't get with anything better instead of) link an external from Tailacale DNS server, so users could connect thru it to the services inside the external infrastructure

**Comment by yegle** [comment]
Note: the following is from a long thread I had with Gemini. I asked Gemini to summarize the thread, manually reviewed/edited, and confirmed it's not a slop.

**Use Case: Subdomain-based Reverse Proxying over Node Sharing**

I am adding my support to this feature request because the lack of custom MagicDNS records (specifically node aliases/CNAMEs) creates a severe architectural roadblock when sharing a multi-service self-hosted node with external users via Node Sharing.
#### The Setup
I host multiple web applications on a single server, multiplexed behind a reverse proxy (Caddy). I share this single node out to my friends. I want them to access services via distinct subdomains rather than cluttering the configuration with distinct network ports or fragile path-based routing (which frequently breaks absolute asset paths in third-party applications).

#### Why Current Workarounds Completely Fail:
 1. **Static Public A Records to 100.x.y.z Are Deprecated:** Now that Tailscale performs localized 1-1 NAT mappings for external shared nodes to prevent pool exhaustion, the node’s IPv4 address is unique to *each recipient's tailnet*. A static public A record pointing to a single CGNAT IP is no longer a viable solution. *Note: AAAA record doesn't solve the problem as it would fail to resolve in an IPv4 only network, despite the resolved IPv6 address is accessible via Tailscale*
 2. **Public CNAME / DNAME Records Hit the OS Resolver Trap:** Pointing a public custom domain (app.domain.com CNAME node.tailnet.ts.net) works on Linux/macOS but completely breaks on Windows and Android. Because public recursive resolvers (1.1.1.1, 8.8.8.8) cannot resolve the .ts.net target, they append an NXDOMAIN status code. The stub resolvers on Windows/Android immediately drop the query upon receiving the public NXDOMAIN before handing the resolution back down to the local Tailscale interface (100.100.100.100).

#### How Native MagicDNS Aliases Resolve This
If MagicDNS allowed us to define arbitrary custom aliases or split-DNS CNAME records inside our managed .ts.net zones—and crucially, **propagated these aliases across Node Sharing boundaries**—external users could cleanly query service1.node.tailnet.ts.net or service2.node.tailnet.ts.net.
Because the query stays entirely within the secure Tailscale layer, it bypasses the public upstream NXDOMAIN trap entirely, maps seamlessly to the recipient's localized 1-1 NAT IP, and allows for clean subdomain multiplexing on a single shared device.

**Comment by Kyobinoyo** [comment]
I also hope this feature makes it into MagicDNS, i don't have a big company but my homelab would be working so much easier with this.
